R/generateIIASASubmission.R

```diff
@@ -62,90 +54,75 @@ generateIIASASubmission <- function(mifs = ".", mapping = NULL, model = "REMIND
-    if (!generateSingleOutput) {
-      outputMif <- file.path(outputDirectory, paste0(outputPrefix, basename(flist[fl])))
-      allmifdata <- NULL
-    }
-    message("# read ", flist[fl])
-    mifdata <- quitte::as.quitte(flist[fl])
-    # remove -rem-xx and mag-xx from scenario names
-    mifdata$scenario <- gsub("-(rem|mag)-[0-9]{1,2}", "", mifdata$scenario)
-    allmifdata <- rbind(allmifdata, mifdata)
-    if (!generateSingleOutput || fl == length(flist)) {
-      message("# Convert to ", outputMif)
-      mifdata <- .setModelAndScenario(allmifdata, model, removeFromScen, addToScen)
-      quitte::write.mif(mifdata, tmpfile)
-      iamc::write.reportProject(
-          tmpfile,
-          mappingFile,
-          file = tmpfile,
-          missing_log = logFile
-      )
-      message("# Restore PM2.5 dot in variable names for consistency with DB template")
-      mifdata <- quitte::as.quitte(tmpfile) %>% mutate(variable = gsub("PM2_5", "PM2.5", !!sym("variable")))
-
-      message("# Replace N/A for missing years with blanks as recommended by Ed Byers")
-      write.mif(mifdata %>% mutate(value = ifelse(is.na(!!sym("value")), "", !!sym("value"))), outputMif)
-
-      mifdata <- mifdata %>%
-        mutate(value = ifelse(!is.finite(!!sym("value")) | is.na(!!sym("value")), NA, !!sym("value"))) %>%
-        mutate(scenario = as.factor(gsub("^NA$", "", !!sym("scenario")))) %>%
-        filter(!!sym("period") %in% timesteps)
-
-      if (!is.null(iiasatemplate) && file.exists(iiasatemplate)) {
-        mifdata <- checkIIASASubmission(mifdata, iiasatemplate, logFile)
-      } else {
-        message("# iiasatemplate ", iiasatemplate, " does not exist, returning full list of variables.")
-      }
+  submission <- mifdata %>%
+    filter(!!sym("period") %in% timesteps) %>%
+    distinct() %>%
+    inner_join(m, by = c("variable" = "piam_variable", "unit" = "piam_unit"), multiple = "all") %>%
+    mutate(
+      !!sym("value") := ifelse(is.na(!!sym("factor")), !!sym("value"), !!sym("factor") * !!sym("value"))
+    ) %>%
+    select("model", "scenario", "region", "period", "variable" = "Variable", "unit" = "Unit", "value")
 
-      # check whether all scenarios have same number of variables
-      countDataPoints <- seq_along(levels(mifdata$scenario))
-      for (i in countDataPoints) {
-        countDataPoints[i] <- sum(mifdata$scenario == levels(mifdata$scenario)[[i]])
-      }
-      if (length(unique(countDataPoints)) != 1) {
-        message(
-          "Not all scenarios have the same data points: ",
-          paste0(levels(mifdata$scenario), ": ", countDataPoints, ".", collapse = " ")
-        )
-      }
 
-      message("# In mif file, replace N/A for missing years with blanks as recommended by Ed Byers")
-      write.mif(mifdata %>% mutate(value = ifelse(is.na(!!sym("value")), "", !!sym("value"))), outputMif)
+  submission <- aggregate(value ~ model + region + scenario + period + variable + unit, data = submission, FUN = "sum")
```

Comment:
I have to admit I don't understand this line.

Reply:
A mapping of two source variables to the same target variable will produce two identical entries that need to be summed up, e.g remindVarA and remindVarB map to ar6VarA, then the two mapped entries must be added